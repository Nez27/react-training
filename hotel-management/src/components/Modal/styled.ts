import styled from 'styled-components';

const StyledModal = styled.div`
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);

  transition: all 0.5s;
`;

const StyledModalContent = styled.div`
  background-color: var(--form-color);

  border-radius: var(--radius-md);

  padding: 30px 30px;
`;

const Overlay = styled.div`
  position: fixed;
  top: 0;
  left: 0;

  width: 100%;
  height: 100vh;

  background-color: var(--overlay-color);

  z-index: 1000;
  transition: all 0.5s;
`;

const TitleModal = styled.p`
  font-size: var(--fs-md);
  font-weight: 600;
  text-align: center;

  margin-bottom: 30px;
`

export { StyledModal, Overlay, StyledModalContent, TitleModal };
