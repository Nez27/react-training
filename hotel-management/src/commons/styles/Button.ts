import styled, { RuleSet, css } from 'styled-components';

interface IVariations {
  [key: string]: RuleSet<object>;
}

const variations: IVariations = {
  primary: css`
    background-color: var(--primary-color);
    color: var(--light-text);
  `,
  secondary: css`
    background-color: var(--secondary-btn-color);
  `,
  danger: css`
    background-color: var(--danger-btn-color);
    color: var(--light-text);
  `,
};

interface IButtonStyle {
  variations?: keyof IVariations;
}

const Button = styled.button<IButtonStyle>`
  padding: 10px 20px;

  text-transform: uppercase;
  font-weight: 600;

  cursor: pointer;

  border-radius: var(--radius-md);

  ${(props) => variations[props.variations!]}

  &:disabled,
  &[disabled] {
    cursor: not-allowed;
    background-color: var(--disabled-btn-color);
  }
`;

Button.defaultProps = {
  variations: 'primary',
};

export default Button;
