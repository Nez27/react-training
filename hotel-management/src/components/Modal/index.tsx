import {
  ReactElement,
  ReactNode,
  cloneElement,
  useContext,
  useState,
} from 'react';
import { createPortal } from 'react-dom';

// Hooks
import { useOutsideClick } from '../../hooks/useOutsideClick';

// Contexts
import { ModalContext } from '../../contexts/ModalContext';

// Styled
import { Overlay, StyledModal, StyledModalContent, TitleModal } from './styled';

interface IModal {
  children: ReactNode;
}

const Modal = ({ children }: IModal) => {
  const [openName, setOpenName] = useState('');

  const close = () => setOpenName('');
  const open = setOpenName;

  return (
    <ModalContext.Provider value={{ openName, close, open }}>
      {children}
    </ModalContext.Provider>
  );
};

interface IOpen {
  children: ReactElement;
  modalName: string;
}

const Open = ({ children, modalName }: IOpen) => {
  const { open } = useContext(ModalContext);

  return cloneElement(children, { onClick: () => open!(modalName) });
};

interface IWindow {
  children: ReactElement;
  name: string;
  title: string;
}

const Window = ({ children, name, title }: IWindow) => {
  const { openName, close } = useContext(ModalContext);
  const ref = useOutsideClick<HTMLDivElement>(close!);

  console.log('name: ', name);
  console.log('openName: ', openName);

  if (name !== openName) return null;

  return createPortal(
    <Overlay>
      <StyledModal ref={ref}>
        <StyledModalContent>
          <TitleModal>{title}</TitleModal>
          {cloneElement(children, { onCloseModal: close })}
        </StyledModalContent>
      </StyledModal>
    </Overlay>,
    document.body
  );
};

Modal.Open = Open;
Modal.Window = Window;

export default Modal;
