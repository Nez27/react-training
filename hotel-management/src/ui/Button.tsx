import styled from 'styled-components';

const Button = styled.button`
  padding: 10px 20px;

  background-color: var(--primary-color);
  color: var(--light-text);
  text-transform: uppercase;
  font-weight: 500;
  cursor: pointer;

  border-radius: var(--radius-sm);
`;

export default Button;
