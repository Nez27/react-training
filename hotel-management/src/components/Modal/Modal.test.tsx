import renderer from 'react-test-renderer';

// Components
import Modal from '.';
import Button from '@commonStyle/Button';

describe('Modal', () => {
  const wrapper = renderer.create(
    <Modal>
      <Modal.Open
        modalName="room-form"
        renderChildren={(onCloseModal) => (
          <Button onClick={onCloseModal}>Add room</Button>
        )}
      />
      <Modal.Window name="room-form" title="Add form">
        <p>This is a windows</p>
      </Modal.Window>
    </Modal>
  );

  test('Should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
