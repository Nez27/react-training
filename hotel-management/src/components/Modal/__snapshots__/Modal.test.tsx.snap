// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Should render correctly 1`] = `
<button
  className="sc-eCAqax iNUTro"
  onClick={[Function]}
  variations="primary"
>
  Add room
</button>
`;
